src/tools/crossgen2/crossgen2/Program.cs

```diff
@@ -163,182 +159,119 @@ private int Run()
 
             using (PerfEventSource.StartStopEvents.CompilationEvents())
             {
-                ICompilation compilation;
+                Queue<ICompilation> compilations = new Queue<ICompilation>();
```

Comment:
Multiple inputs, multiple output modes tend to be more difficult to hook into build (problems with incrementality, degree of parallelism, memory use, warning generation and attribution to a specific file, compiler crash diagnostic, etc.)

What advantage are we getting out of this, compared to re-invoking the tool from outside?

Reply:
I was under the impression that, at some point, we'll also add the variant "multiple inputs, single output". In such case this might be a reasonable preparatory change.

Reply:
I would expect "multiple inputs, single output" to have only a single `Compilation` object - each compilation produces an independent dependency graph - it feels like having multiple independent graphs would just make things harder.

Reply:
Hmm, you're probably right. I think that the "multiple dependency graphs" version popped up in our composite / single file discussions at some point but is seems much less interesting and useful than the "fully cross-optimized" composite output.

Reply:
We talked about allowing multiple inputs at an AOT sync a month or so back, speculating it might yield publish throughput improvements by not needing to invoke Crossgen2 multiple times and eat the costs associated with that. It also makes the tool friendlier for the dev team so we can compile a number of assemblies easily without relying on SuperIlc (which exists because it's hard to use Crossgen2 today outside of a build system).

I agree, this would not work well with the build system so I see this as used by the runtime devs, and for compiling tests more easily.

If folks don't want to go this direction, I could keep the refactor portions and leave out the multi-in piece.

Reply:
Well, from my perspective I don't have any problems with that. But then it was me who implemented a similar thing in .NET Native ILC long ago and Michal also didn't like that :-).

Reply:
> publish throughput improvements by not needing to invoke Crossgen2 multiple times

That's why we're trying to improve startup time in .NET 5. Sounds like more motivation to do better there. We can start with running crossgen2 on itself.

> I see this as used by the runtime devs, and for compiling tests more easily

Would checking in a batch/shell script that runs crossgen2 in a loop on all files in a directory help alleviate that? We could even binplace it next to crossgen2.

I'm not opposed to test hooks - I'm just trying to be careful about test hooks that end up shipping with the product.